test/cpu_adaptiveavgpool.hpp

```diff
@@ -0,0 +1,311 @@
+
+        size_t h  = static_cast<size_t>(std::floor(static_cast<float>(oh * H) / OH));
+        size_t kh = static_cast<size_t>(std::ceil(static_cast<float>((oh + 1) * H) / OH)) - h;
+
+        size_t w  = static_cast<size_t>(std::floor(static_cast<float>(ow * W) / OW));
+        size_t kw = static_cast<size_t>(std::ceil(static_cast<float>((ow + 1) * W) / OW)) - w;
+
+        float sum = 0;
+        for(size_t id = d; id < (d + kd); ++id)
+        {
+            for(size_t ih = h; ih < (h + kh); ++ih)
+            {
+                for(size_t iw = w; iw < (w + kw); ++iw)
+                {
+                    sum +=
+                        static_cast<float>(input[input_tv.get_tensor_view_idx({n, c, id, ih, iw})]);
+                }
+            }
+        }
+
+        output[output_tv.get_tensor_view_idx({n, c, od, oh, ow})] =
+            static_cast<T>(sum / (kd * kh * kw));
+    }
+}
+
+template <class T>
+void cpu_adaptiveavgpool_backward_1d(
+    tensor<T> output_grad, tensor<T>& input_grad, size_t N, size_t C, size_t H, size_t OH)
+{
+    auto dims  = input_grad.desc.GetLengths();
+    auto numel = input_grad.desc.GetElementSize();
+
+    auto output_grad_tv = miopen::get_inner_expanded_tv<3>(output_grad.desc);
+    auto input_grad_tv  = miopen::get_inner_expanded_tv<3>(input_grad.desc);
+
+    for(size_t gid = 0; gid < numel; gid++)
+    {
+        size_t nc = gid / H, h = gid % H;
+        size_t n = nc / C, c = nc % C;
+
+        if(n >= N)
+            return;
+
+        size_t oh  = static_cast<size_t>(std::floor(static_cast<float>(h * OH) / H));
+        size_t koh = static_cast<size_t>(std::ceil(static_cast<float>((h + 1) * OH) / H)) - oh;
+
+        float grad = 0;
+        for(size_t ih = oh; ih < (oh + koh); ++ih)
+        {
+            size_t kh = static_cast<size_t>(std::ceil(static_cast<float>((ih + 1) * H) / OH)) -
+                        static_cast<size_t>(std::floor(static_cast<float>(ih * H) / OH));
+            grad +=
+                static_cast<float>(output_grad[output_grad_tv.get_tensor_view_idx({n, c, ih})]) /
+                kh;
+        }
+
+        input_grad[input_grad_tv.get_tensor_view_idx({n, c, h})] = static_cast<T>(grad);
+    }
+}
+
+template <class T>
+void cpu_adaptiveavgpool_backward_2d(tensor<T> output_grad,
+                                     tensor<T>& input_grad,
+                                     size_t N,
+                                     size_t C,
+                                     size_t H,
+                                     size_t W,
+                                     size_t OH,
+                                     size_t OW)
+{
+    auto dims  = input_grad.desc.GetLengths();
+    auto numel = input_grad.desc.GetElementSize();
+
+    auto output_grad_tv = miopen::get_inner_expanded_tv<4>(output_grad.desc);
+    auto input_grad_tv  = miopen::get_inner_expanded_tv<4>(input_grad.desc);
+
+    for(size_t gid = 0; gid < numel; gid++)
```

Comment:
You should use `par_ford` to perform parallel calculation, for better performance.